### PR TITLE
Send and receive CloseSession frame

### PR DESCRIPTION
**Pull request by ddragana**

### CODE REVIEW

**Comment by ddragana** [comment]
I want to explore a different architecture for WebTransport CONNECT stream.

I do not like that Http3Connection (connection.rs) needs to know about WebTransport CONNECT streams to be able to manage them. 

I am planing to try creating a special WebTransport stream that holds RecvMessage and SendMessage. With this design the WebTransport stream would be put into the Http3Connection's send_streams and recv_streams and they will manage their own live time. so we do not need a hashmap of WebTransport sessions in ExtendedConnectFeature. Also Http3Connection does not need a special handling for WebTransport streams.

**Comment by ddragana** [comment]
One more think that is missing is that CloseSession frame needs to be send whole. The current implementation does not guarantee that.

**Comment by ddragana** [comment]
This also needs more tests.

**Comment by ddragana** [comment]
This fixes #1286

File: neqo-http3/src/features/extended_connect/webtransport_session.rs [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+            frame_reader: FrameReader::new(),
+            events,
+            send_streams: BTreeSet::new(),
+            recv_streams: BTreeSet::new(),
+            role,
+        }
+    }
+
+    /// # Errors
+    /// The above function can only fail if supplied headers are not valid http headers.
```
Comment (KershawChang) [inline neqo-http3/src/features/extended_connect/webtransport_session.rs:137]: nit: which `above` function? I think you mean `send_request`, right?

File: neqo-http3/src/features/extended_connect/webtransport_session.rs [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+    fn has_data_to_send(&self) -> bool {
+        self.control_stream_send.has_data_to_send()
+    }
+
+    fn done(&self) -> bool {
+        self.state == SessionState::Done
+    }
+
+    fn handle_stop_sending(&mut self, close_type: CloseType) {
+        self.close(close_type);
```
Comment (KershawChang) [inline neqo-http3/src/features/extended_connect/webtransport_session.rs:211]: This function does exactly the same as `reset`. Maybe we can remove one of them.

File: neqo-http3/src/connection_client.rs [hunk 3]
```diff
@@ -436,6 +435,22 @@ impl Http3Client {
         output
     }
 
+    /// Close `WebTransport` cleanly
+    /// # Errors
+    /// `InvalidStreamId` if thee stream does not exist,
```
Comment (KershawChang) [inline neqo-http3/src/connection_client.rs:440]: nit: `the`

**Review by KershawChang — APPROVED** [review]
I have some nits and minor comments, but I think this is good to land.

File: neqo-http3/src/recv_message.rs [hunk 4]
```diff
@@ -468,15 +473,30 @@ impl HttpRecvStream for RecvMessage {
...
+    fn priority_update_frame(&mut self) -> Option<HFrame> {
+        self.priority_handler.maybe_encode_frame(self.stream_id)
+    }
+
+    fn priority_update_sent(&mut self) {
+        self.priority_handler.priority_update_sent();
     }
 
     fn set_new_listener(&mut self, conn_events: Box<dyn HttpRecvStreamEvents>) {
+        self.state = RecvMessageState::WaitingForData {
```
[the remark this answers is not in the document]
Comment (ddragana) [reply neqo-http3/src/recv_message.rs:489]: I am planning to refactor this a bit. My idea is: when the code figures out that this is a CONNECT request it should transform it into a ExtendedConnect stream. This is not easy to do in rust therefore I have not doe int in this change.

I will leave it for now.

File: neqo-http3/src/send_message.rs [hunk 5]
```diff
@@ -306,8 +320,13 @@ impl HttpSendStream for SendMessage {
     }
 
     fn set_new_listener(&mut self, conn_events: Box<dyn SendStreamEvents>) {
+        self.stream_type = Http3StreamType::ExtendedConnect;
```
[the remark this answers is not in the document]
Comment (ddragana) [reply neqo-http3/src/send_message.rs:323]: This  will part of the same refactoring as above.

File: neqo-http3/src/features/extended_connect/webtransport_session.rs (hunk 2 above)
Comment (ddragana) [reply neqo-http3/src/features/extended_connect/webtransport_session.rs:211]: Nice catch, both of them can be remove. They only call close().

**Pull request merged**